subsys/mgmt/ec_host_cmd/ec_host_cmd_handler.c

```diff
@@ -133,79 +180,56 @@ static void handle_host_cmds_entry(void *arg1, void *arg2, void *arg3)
-					    EC_HOST_CMD_INVALID_COMMAND);
+			send_error_response(hc->backend, tx, EC_HOST_CMD_INVALID_COMMAND);
 			continue;
 		}
 
-		struct ec_host_cmd_handler_args args = {
-			.reserved = NULL,
-			.command = rx_header->cmd_id,
-			.version = rx_header->cmd_ver,
-			.input_buf = rx.buf + RX_HEADER_SIZE,
-			.input_buf_size = rx_header->data_len,
-			.output_buf = tx_buffer + TX_HEADER_SIZE,
-			.output_buf_max = sizeof(tx_buffer) - TX_HEADER_SIZE,
-			.output_buf_size = 0,
-		};
-
-		if (found_handler->min_rqt_size > args.input_buf_size) {
-			send_error_response(ec_host_cmd_dev,
-					    EC_HOST_CMD_REQUEST_TRUNCATED);
-			continue;
-		}
+		args.command = rx_header->cmd_id;
+		args.version = rx_header->cmd_ver;
+		args.input_buf_size = rx_header->data_len;
+		args.output_buf_size = 0;
 
-		if (found_handler->min_rsp_size > args.output_buf_max) {
-			send_error_response(ec_host_cmd_dev,
-					    EC_HOST_CMD_INVALID_RESPONSE);
+		status = validate_handler(found_handler, &args);
+		if (status != EC_HOST_CMD_SUCCESS) {
+			send_error_response(hc->backend, tx, status);
 			continue;
 		}
 
-		if (args.version > sizeof(found_handler->version_mask) ||
-		    !(found_handler->version_mask & BIT(args.version))) {
-			send_error_response(ec_host_cmd_dev,
-					    EC_HOST_CMD_INVALID_VERSION);
+		status = found_handler->handler(&args);
+		if (status != EC_HOST_CMD_SUCCESS) {
+			send_error_response(hc->backend, tx, status);
 			continue;
 		}
 
-		const enum ec_host_cmd_status handler_rv =
-			found_handler->handler(&args);
-
-		if (handler_rv != EC_HOST_CMD_SUCCESS) {
-			send_error_response(ec_host_cmd_dev, handler_rv);
+		status = prepare_response(tx, args.output_buf_size);
+		if (status != EC_HOST_CMD_SUCCESS) {
+			send_error_response(hc->backend, tx, status);
 			continue;
 		}
 
-		struct ec_host_cmd_response_header *const tx_header =
-			(void *)tx_buffer;
-
-		tx_header->prtcl_ver = 3;
-		tx_header->result = EC_HOST_CMD_SUCCESS;
-		tx_header->data_len = args.output_buf_size;
-		tx_header->reserved = 0;
+		hc->backend->api->send(hc->backend);
+	}
+}
 
-		const uint16_t tx_valid_data_size =
-			tx_header->data_len + TX_HEADER_SIZE;
-		if (tx_valid_data_size > sizeof(tx_buffer)) {
-			send_error_response(ec_host_cmd_dev,
-					    EC_HOST_CMD_INVALID_RESPONSE);
-			continue;
-		}
+int ec_host_cmd_init(struct ec_host_cmd_backend *backend, const void *backend_config)
+{
+	struct ec_host_cmd *hc = &ec_host_cmd;
+	int ret;
 
-		/* Calculate checksum */
-		tx_header->checksum = 0;
-		tx_header->checksum =
-			cal_checksum(tx_buffer, tx_valid_data_size);
+	hc->backend = backend;
 
-		const struct ec_host_cmd_tx_buf tx = {
-			.buf = tx_buffer,
-			.len = tx_valid_data_size,
-		};
+	/* Allow writing to rx buff at startup */
+	k_sem_init(&hc->rx_ctx.handler_owns, 0, 1);
 
-		ec_host_cmd_backend_send(ec_host_cmd_dev, &tx);
+	ret = backend->api->init(backend, backend_config, &hc->rx_ctx, &hc->tx);
+	if (ret != 0) {
```

Comment:
Verify that the buffers are not NULL. Also check if the buffers changed and log a warning so the user can eliminate the duplicate buffer.